File: app/src/main/java/seigneur/gauvain/mycourt/ui/shots/ShotsViewModel.java
package seigneur.gauvain.mycourt.ui.shots;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import seigneur.gauvain.mycourt.data.model.Shot;
import seigneur.gauvain.mycourt.data.repository.ShotRepository;
import seigneur.gauvain.mycourt.data.repository.TempDataRepository;
import seigneur.gauvain.mycourt.ui.shots.list.data.NetworkState;
import seigneur.gauvain.mycourt.ui.shots.list.data.datasource.ShotDataSourceFactory;
import seigneur.gauvain.mycourt.ui.shots.list.data.datasource.ShotsDataSource;
import seigneur.gauvain.mycourt.utils.ConnectivityReceiver;
import seigneur.gauvain.mycourt.utils.SingleLiveEvent;
import seigneur.gauvain.mycourt.utils.rx.NetworkErrorHandler;
import timber.log.Timber;


public class ShotsViewModel extends ViewModel {

    @Inject
    NetworkErrorHandler mNetworkErrorHandler;

    @Inject
    ConnectivityReceiver mConnectivityReceiver;

    @Inject
    TempDataRepository mTempDataRepository;

    @Inject
    ShotRepository mShotRepository;

    LiveData<PagedList<Shot>> shotList;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private static final int pageSize = 15;

    private ShotDataSourceFactory shotDataSourceFactory;

    private PagedList.Config config;

    private SingleLiveEvent<Integer> shotClickEvent = new SingleLiveEvent<>();

    @Inject
    public ShotsViewModel() { }

    public void init() {
        if (shotDataSourceFactory==null && config==null && shotList==null) {
            shotDataSourceFactory = new
                    ShotDataSourceFactory(compositeDisposable, mShotRepository);
            config = new PagedList.Config.Builder()
                    .setPageSize(pageSize)
                    .setInitialLoadSizeHint(pageSize)
                    .setEnablePlaceholders(false)
                    .build();
            shotList = new LivePagedListBuilder<>(shotDataSourceFactory, config).build();
        }

    }

    public void retry() {
        if (shotDataSourceFactory.getUsersDataSourceLiveData().getValue()!=null)
            shotDataSourceFactory.getUsersDataSourceLiveData().getValue().retry();
    }

    public void refresh() {
        if (shotDataSourceFactory.getUsersDataSourceLiveData().getValue()!=null)
            shotDataSourceFactory.getUsersDataSourceLiveData().getValue().invalidate();
    }

    public LiveData<NetworkState> getNetworkState() {
        return Transformations.switchMap(shotDataSourceFactory.getUsersDataSourceLiveData(),
                ShotsDataSource::getNetworkState);
    }

    public LiveData<NetworkState> getRefreshState() {
        return Transformations.switchMap(shotDataSourceFactory.getUsersDataSourceLiveData(),
                ShotsDataSource::getInitialLoad);
    }

    public SingleLiveEvent<Integer> getShotClickEvent() {
        return shotClickEvent;
    }

    public void onShotClicked(Shot shot, int position) {
        shotClickEvent.setValue(position);
        mTempDataRepository.setShot(shot);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.dispose();
    }



}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/shots/ShotsViewModel.kt
package seigneur.gauvain.mycourt.ui.shots

import androidx.arch.core.util.Function
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList

import javax.inject.Inject

import io.reactivex.disposables.CompositeDisposable
import seigneur.gauvain.mycourt.data.model.Shot
import seigneur.gauvain.mycourt.data.repository.ShotRepository
import seigneur.gauvain.mycourt.data.repository.TempDataRepository
import seigneur.gauvain.mycourt.ui.shots.list.data.NetworkState
import seigneur.gauvain.mycourt.ui.shots.list.data.datasource.ShotDataSourceFactory
import seigneur.gauvain.mycourt.ui.shots.list.data.datasource.ShotsDataSource
import seigneur.gauvain.mycourt.utils.ConnectivityReceiver
import seigneur.gauvain.mycourt.utils.SingleLiveEvent
import seigneur.gauvain.mycourt.utils.rx.NetworkErrorHandler
import timber.log.Timber


class ShotsViewModel @Inject
constructor() : ViewModel() {

    @Inject
    lateinit var mNetworkErrorHandler: NetworkErrorHandler

    @Inject
    lateinit var mConnectivityReceiver: ConnectivityReceiver

    @Inject
    lateinit var mTempDataRepository: TempDataRepository

    @Inject
    lateinit var mShotRepository: ShotRepository

    var shotList: LiveData<PagedList<Shot>>? = null

    private val compositeDisposable = CompositeDisposable()

    private val shotDataSourceFactory: ShotDataSourceFactory by lazy {
        ShotDataSourceFactory(compositeDisposable, mShotRepository)
    }

    private var config: PagedList.Config? = null

    val shotClickEvent = SingleLiveEvent<Int>()

    val networkState: LiveData<NetworkState>
        get() =  Transformations.switchMap(shotDataSourceFactory.usersDataSourceLiveData)
        { it.networkState }

    val refreshState: LiveData<NetworkState>
        get() = Transformations.switchMap(shotDataSourceFactory.usersDataSourceLiveData) {
            Timber.d("refresh called ")
            it.initialLoad
        }


    fun init() {
        if (config == null && shotList == null) {
            config = PagedList.Config.Builder()
                    .setPageSize(pageSize)
                    .setInitialLoadSizeHint(pageSize)
                    .setEnablePlaceholders(false)
                    .build()
            shotList = LivePagedListBuilder(shotDataSourceFactory, config!!).build()
        }

    }

    fun retry() {
        if (shotDataSourceFactory.usersDataSourceLiveData.value != null)
            shotDataSourceFactory.usersDataSourceLiveData.value!!.retry()
    }

    fun refresh() {
        if (shotDataSourceFactory.usersDataSourceLiveData.value != null)
            shotDataSourceFactory.usersDataSourceLiveData.value!!.invalidate()
    }

    fun onShotClicked(shot: Shot, position: Int) {
        shotClickEvent.value = position
        mTempDataRepository.shot = shot
    }

    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }

    companion object {

        private val pageSize = 15
    }


}
